types/ejs/index.d.ts

```diff
@@ -28,14 +28,22 @@ export function resolveInclude(name: string, filename: string, isDir: boolean):
 /**
  * Compile the given `str` of ejs into a template function.
  */
-export function compile(template: string, opts?: Options): (TemplateFunction);
+export function compile(template: string, opts?: Options & { async: false }): TemplateFunction;
+export function compile(template: string, opts: Options & { async: true }): AsyncTemplateFunction;
+export function compile(template: string, opts: Options & { async: boolean }): TemplateFunction | AsyncTemplateFunction;
+export function compile(template: string, opts: Exclude<Options, { async: any }>): TemplateFunction;
+export function compile(template: string, opts?: Options): TemplateFunction | AsyncTemplateFunction;
 /**
  * Render the given `template` of ejs.
  *
  * If you would like to include options but not data, you need to explicitly
  * call this function with `data` being an empty object or `null`.
  */
-export function render(template: string, data?: Data, opts?: Options): string;
+export function render(template: string, data?: Data, opts?: Options & { async: false }): string;
+export function render(template: string, data: Data, opts: Options & { async: true }): Promise<string>;
+export function render(template: string, data: Data, opts: Options & { async: boolean }): string | Promise<string>;
+export function render(template: string, data: Data, opts: Exclude<Options, { async: any }>): string;
+export function render(template: string, data?: Data, opts?: Options): string | Promise<string>;
```

Comment:
Here, you have more issues than above; as it stands currently
```
const result1 = ejs.render(template, undefined, { async: true });
```
gives the type `string | Promise<string>` for `result1`, which if I understand correctly, should not hold.

You can simply fix this error like the "don't use TS 2.8" case above by making `data` optional always and remove the `ExcludeOptions` usage as before.
Or you can use conditional types directly again as follows:
```
export function render<T extends Options | undefined>(template: string, data: Data | undefined, opts: T): T extends { async: true }
    ? Promise<string>
    : T extends { async: false }
        ? string
        : T extends { async?: boolean }
            ? string | Promise<string>
            : string;
export function render(template: string, data?: Data): string;
```

Reply:
As above, rather than using the conditional type, intersecting with `{ async?: never }` will give the desired result.